### PR TITLE
Bugfix/#48 new authentication flow

### LINKED ISSUE #48: Authentification no longer working
From this day, 14th of September 2019, trying to log in with this client, whether from the module or with the CLI application, will result in a 403 error:

```
$ n26 balance
Traceback (most recent call last):
  File "/home/geoffrey/.local/bin/n26", line 11, in <module>
    load_entry_point('n26==1.2.0', 'console_scripts', 'n26')()
  File "/usr/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/geoffrey/.local/lib/python3.7/site-packages/n26/cli.py", line 94, in balance
    balance_data = API_CLIENT.get_balance()
  File "/home/geoffrey/.local/lib/python3.7/site-packages/n26/api.py", line 59, in get_balance
    return self._do_request(GET, BASE_URL + '/api/accounts')
  File "/home/geoffrey/.local/lib/python3.7/site-packages/n26/api.py", line 214, in _do_request
    access_token = self.get_token()
  File "/home/geoffrey/.local/lib/python3.7/site-packages/n26/api.py", line 272, in get_token
    self._token_data = self._request_token(self.config.username, self.config.password)
  File "/home/geoffrey/.local/lib/python3.7/site-packages/n26/api.py", line 296, in _request_token
    response.raise_for_status()
  File "/usr/lib/python3.7/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://api.tech26.de/oauth/token
```

I believe this is due to enforcement of the Revised Directive on Payment Services (PSD2) enforcing "strong" authentication. In fact, from now on, connecting to the N26 web application requires you to validate the login attempt from the phone application.

So I guess the big question is, can this be worked around, or should we say goodbye to the N26 API (without an validation from the phone application, which defeats the purpose)?

### PR DESCRIPTION
**Pull request by markusressel**
fixes #48 

This PR implements the new 2FA required by N26.

To prevent having to reapprove every single CLI command a new config option has been added to specify a file path where access token data can be saved on the file system. This config option is optional though and by default `None`. If this config option is not set the token data will only be stored in memory and never written to disk.

After issuing a command without a previous authentication attempt the user has 60 seconds to approve the login using his paired phone and N26 app. To implement a simple "retry" logic I added the `tenacity` dependency.

### CODE REVIEW

File: n26/api.py [hunk 1]
```diff
@@ -229,33 +273,6 @@ def _do_request(self, method: str = GET, url: str = "/", params: dict = None,
         if len(response.content) > 0:
             return response.json()
 
-    @staticmethod
-    def _create_request_url(url: str, params: dict = None):
```
Comment (markusressel) [inline n26/api.py]: This method has moved unaltered to `util.py`

File: n26/api.py [hunk 2]
```diff
@@ -265,37 +282,51 @@ def get_token(self):
...
-            'password': password
+            "grant_type": GRANT_TYPE_PASSWORD,
+            "username": username,
+            "password": password
         }
-
-        response = requests.post(BASE_URL + '/oauth/token', data=values_token, headers=BASIC_AUTH_HEADERS)
-        response.raise_for_status()
-        return response.json()
+        # TODO: Seems like the user-agent is not necessary but might be a good idea anyway
```
Comment (markusressel) [inline n26/api.py]: I'm not sure if we want this, I don't think it is needed, at least not at the moment.

Comment (markusressel) [reply n26/api.py]: I will leave this TODO in the code for now, we can think about it again if it becomes a problem.

**Comment by markusressel** [comment]
Although the mixed use of the `.de` and `.global` domains is probably irrelevant I dont want to waste another day to thoroughly test this so I will just leave it in the current state. I will open an issue for it and we can clean this up at a later stage.

**Comment by SoftCreatR** [comment]
Is there a way to refresh the token after expiry (899 seconds) using the `refresh_token`?

**Comment by markusressel** [comment]
Yes it is. From my understanding thats the way it is supposed to be and how it is implemented in this library. We dont know exactly how long it is possible though since this is not specified in the jwt specification and is hard to test as well.

**Comment by SoftCreatR** [comment]
Okay then. Seems, that the token doesn't hold longer than a day (i've tested your library yesterday and today, i had to re-log-in).

**Comment by markusressel** [comment]
If you want to keep the login I recommend you execute some call at least every 30 minutes or so. This should keep the token valid forever.

**Comment by SoftCreatR** [comment]
Makes absolute sense. Ty for the hint.

**Pull request merged**